src/number_theory/diophantine_approximation.lean

```diff
@@ -0,0 +1,363 @@
+    refine ⟨⌊ξ * m⌋ + 1, m, hm₀, (mem_Icc.mp hm).2, _⟩,
+    rw [cast_add, ← sub_sub, sub_mul, cast_one, one_mul, abs_le],
+    refine ⟨le_sub_iff_add_le.mpr _, sub_le_iff_le_add.mpr $ le_of_lt $
+             (hfu m).trans $ lt_one_add _⟩,
+    simpa only [neg_add_cancel_comm_assoc] using hf', },
+  { simp_rw [not_exists] at H,
+    have hD : (Ico (0 : ℤ) n).card < D.card,
+    { rw [card_Icc, card_Ico], exact lt_add_one n, },
+    have hfu' : ∀ m, f m ≤ n := λ m, lt_add_one_iff.mp (floor_lt.mpr (by exact_mod_cast hfu m)),
+    have hwd : ∀ m : ℤ, m ∈ D → f m ∈ Ico (0 : ℤ) n :=
+      λ x hx, mem_Ico.mpr ⟨floor_nonneg.mpr (mul_nonneg (fract_nonneg (ξ * x)) hn.le),
+                           ne.lt_of_le (H x hx) (hfu' x)⟩,
+    have : ∃ (x : ℤ) (hx : x ∈ D) (y : ℤ) (hy : y ∈ D), x < y ∧ f x = f y,
+    { obtain ⟨x, hx, y, hy, x_ne_y, hxy⟩ := exists_ne_map_eq_of_card_lt_of_maps_to hD hwd,
+      rcases lt_trichotomy x y with h | h | h,
+      exacts [⟨x, hx, y, hy, h, hxy⟩, false.elim (x_ne_y h), ⟨y, hy, x, hx, h, hxy.symm⟩], },
+    obtain ⟨x, hx, y, hy, x_lt_y, hxy⟩ := this,
+    refine ⟨⌊ξ * y⌋ - ⌊ξ * x⌋, y - x, sub_pos_of_lt x_lt_y,
+            sub_le_iff_le_add.mpr $ le_add_of_le_of_nonneg (mem_Icc.mp hy).2 (mem_Icc.mp hx).1, _⟩,
+    convert_to |fract (ξ * y) * (n + 1) - fract (ξ * x) * (n + 1)| ≤ 1,
+    { congr, push_cast, simp only [fract], ring, },
+    exact (abs_sub_lt_one_of_floor_eq_floor hxy.symm).le, }
+end
+
+/-- *Dirichlet's approximation theorem:*
+For any real number `ξ` and positive natural `n`, there is a fraction `q`
+such that `q.denom ≤ n` and `|ξ - q| ≤ 1/((n+1)*q.denom)`. -/
+lemma dirichlet_approx' (ξ : ℝ) {n : ℕ} (n_pos : 0 < n) :
+  ∃ q : ℚ, |ξ - q| ≤ 1 / ((n + 1) * q.denom) ∧ q.denom ≤ n :=
+begin
+  obtain ⟨j, k, hk₀, hk₁, h⟩ := dirichlet_approx ξ n_pos,
+  have hk₀' : (0 : ℝ) < k := int.cast_pos.mpr hk₀,
+  have hden : ((j / k : ℚ).denom : ℤ) ≤ k,
+  { convert le_of_dvd hk₀ (rat.denom_dvd j k), exact rat.coe_int_div_eq_mk, },
+  refine ⟨j / k, _, nat.cast_le.mp (hden.trans hk₁)⟩,
+  rw [← div_div, le_div_iff (nat.cast_pos.mpr $ rat.pos _ : (0 : ℝ) < _)],
+  refine (mul_le_mul_of_nonneg_left (int.cast_le.mpr hden : _ ≤ (k : ℝ)) (abs_nonneg _)).trans _,
+  rwa [← abs_of_pos hk₀', rat.cast_div, rat.cast_coe_int, rat.cast_coe_int,
+       ← abs_mul, sub_mul, div_mul_cancel _ hk₀'.ne'],
+end
+
+end pigeonhole
+
+section rat_approx
+
+/-!
+### Equivalence between `rat_approx ξ` and approximating fractions
+
+We define `dioph_approx.rat_approx ξ` and show that it is bijective to the
+set of fractions `q` such that `|ξ - q| < 1/q.denom^2`. We also prove some
+more properties.
+-/
+
+open set
+
+/-- We define the set `rat_approx ξ` of good rational approximations to `ξ`
+as a set of pairs `(x, y)` of integers (corresponding to the fraction `x/y`). -/
+def rat_approx (ξ : ℝ) : set (ℤ × ℤ) :=
+  {xy : ℤ × ℤ | 0 < xy.2 ∧ int.gcd xy.1 xy.2 = 1 ∧ |ξ - xy.1 / xy.2| < 1 / xy.2 ^ 2}
```

Comment:
I would expect the main definition to be a set of type `set ℚ` rather than `set (ℤ × ℤ)`. I would hope that we have sufficient API for `ℚ` to render it unnecessary to work in terms of raw pairs of integers. However if not, I think the main definition should have type `set ℚ`. I also note in passing that when I'd expect `ℤ × ℕ` (non-negative denominator) would be more convenient than `ℤ × ℤ`.

Also, this looks it is worth generalising to allow a general distance function rather than just `m/n ↦ 1/n^2`.

Finally I think that `rat_approx` is too short for the global namespace.

I thus suggest:
```suggestion
/-- Given `ξ : ℝ` and a distance function `f : ℚ → ℝ`, `ξ.approx_rat f` is the set of rational
numbers `q` such that `ξ` is within a distance `f q` of `q`. -/
def real.approx_rat (ξ : ℝ) (f : ℚ → ℝ) : set ℚ := {q | |ξ - q| < f q}
```

Reply:
Again, we are in the namespace `dioph_approx` here.
`rat_approx` is used as a convenient abbreviation for the set of pairs of integers such that etc. The version with rational numbers is spelled out explicitly as it is not that long (see, e.g., `rat_approx_infinite` om line 283-284). I want to have versions of the statements with pairs of integers (BTW, I don't see any upside in using `ℤ × ℕ` over `ℤ × ℤ` -- integers are better behaved, and one would not save the extra condition "denominator positive" (or nonzero)), since this is useful for applications (e.g., to Pell's equation).
The version you suggest will certainly be useful elsewhere (in contexts like what you did on well-approximable elements), but here it would be cumbersome to always carry along `f`, which would be the same function throughout.

Reply:
> The version with rational numbers is spelled out explicitly as it is not that long (see, e.g., `rat_approx_infinite` om line 283-284). 

I am aware of this but I care much more about definitions than lemmas. Of course we can make any definition work by supplying sufficient API but there are still better and worse definitions. My hope is that by working with `set ℚ` we could lean on the API for `ℚ` and so need less boilerplate code overall.

> I don't see any upside in using `ℤ × ℕ` over `ℤ × ℤ` -- integers are better behaved, and one would not save the extra condition "denominator positive" (or nonzero),

The advantages I expect to enjoy from using `ℤ × ℕ` instead of `ℤ × ℤ` are:

 1. There is less redundancy of representation since we avoid the `x / y = (-x) / (-y)` problem.
 2. Our implementation of `ℚ` (essentially) uses `ℤ × ℕ` so it should be easier to connect with `ℚ`.

> I want to have versions of the statements with pairs of integers since this is useful for applications (e.g., to Pell's equation).

It is hard to weigh this point. I think we should be able to judge this code on its own merits.

> it would be cumbersome to always carry along `f`, which would be the same function throughout.

Usually when I make a generalisation like this, I find that several of my lemmas apply more generally and so it is worth it. However if you do not find this is the case then we can just drop this generalisation.

Reply:
It looks like our comments were done in parallel ...

Reply:
> > The version with rational numbers is spelled out explicitly as it is not that long (see, e.g., `rat_approx_infinite` om line 283-284).
> 
> I am aware of this but I care much more about definitions than lemmas. Of course we can make any definition work by supplying sufficient API but there are still better and worse definitions. My hope is that by working with `set ℚ` we could lean on the API for `ℚ` and so need less boilerplate code overall.

I still find it useful to have both versions around (and the statement that shows they are equivalent). For example, the proof of finiteness for rational numbers is a bit easier with the pairs version.
 
> > I don't see any upside in using `ℤ × ℕ` over `ℤ × ℤ` -- integers are better behaved, and one would not save the extra condition "denominator positive" (or nonzero),
> 
> The advantages I expect to enjoy from using `ℤ × ℕ` instead of `ℤ × ℤ` are:
> 
> 1. There is less redundancy of representation since we avoid the `x / y = (-x) / (-y)` problem.
> 
> 2. Our implementation of `ℚ` (essentially) uses `ℤ × ℕ` so it should be easier to connect with `ℚ`.

The redundancy is excluded since the denominator is required to be positive. The connection with `ℚ` is not difficult with the version used here; see `rat_approx_equiv`. I don't think having a natural number as second component makes the proof any easier (in fact, `rat.num_div_eq_of_coprime` etc. work with integers...).

> > it would be cumbersome to always carry along `f`, which would be the same function throughout.
> 
> Usually when I make a generalisation like this, I find that several of my lemmas apply more generally and so it is worth it. However if you do not find this is the case then we can just drop this generalisation.

Everything in this file is quite specific to `1/q.denom^2`, so I don't see how more general results could emerge.